pkg/processor/procbuiltin/unwrap.go

```diff
@@ -0,0 +1,350 @@
+	if p, ok := payload.(map[string]interface{}); ok {
+		structPayload = p
+	} else {
+		return record.Record{}, cerrors.Errorf("payload doesn't contain a record")
+	}
+
+	// check fields under payload
+	err := d.validateRecord(structPayload)
+	if err != nil {
+		return record.Record{}, err
+	}
+
+	var before record.StructuredData
+	beforeData := structPayload[debeziumFieldBefore]
+	b, ok := beforeData.(map[string]any)
+	if beforeData != nil && !ok {
+		return record.Record{}, cerrors.Errorf("%s field is not a map", debeziumFieldBefore)
+	}
+	before = b
+
+	var after record.StructuredData
+	afterData := structPayload[debeziumFieldAfter]
+	a, ok := afterData.(map[string]any)
+	if afterData != nil && !ok {
+		return record.Record{}, cerrors.Errorf("%s field is not a map", debeziumFieldAfter)
+	}
+	after = a
+
+	op, ok := structPayload[debeziumFieldOp].(string)
+	if !ok {
+		return record.Record{}, cerrors.Errorf("%s operation is not a string", op)
+	}
+
+	operation, err := d.convertOperation(op)
+	if err != nil {
+		return record.Record{}, cerrors.Errorf("error unwrapping operation: %w", err)
+	}
+
+	metadata, err := d.unwrapMetadata(rec)
+	if err != nil {
+		return record.Record{}, cerrors.Errorf("error unwrapping metadata: %w", err)
+	}
+
+	// create a kafka connect unwrapper to unwrap the key
+	kafkaConnectUnwrapper := &KafkaConnectUnwrapper{}
+	key := kafkaConnectUnwrapper.UnwrapKey(rec.Key)
+
+	return record.Record{
+		Key:       key,
+		Position:  rec.Position,
+		Operation: operation,
+		Payload: record.Change{
+			Before: before,
+			After:  after,
+		},
+		Metadata: metadata,
+	}, nil
+}
+
+func (d *DebeziumUnwrapper) validateRecord(data record.StructuredData) error {
+	var multiErr error
+	if _, ok := data[debeziumFieldBefore]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", debeziumFieldBefore))
+	}
+	if _, ok := data[debeziumFieldAfter]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", debeziumFieldAfter))
+	}
+	if _, ok := data[debeziumFieldSource]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", debeziumFieldSource))
+	}
+	if _, ok := data[debeziumFieldOp]; !ok {
+		multiErr = multierror.Append(multiErr, cerrors.Errorf("the %q field is missing from debezium payload", debeziumFieldOp))
+	}
+	// ts_ms and transaction can be empty
+	return multiErr
+}
+
+func (d *DebeziumUnwrapper) unwrapMetadata(rec record.Record) (record.Metadata, error) {
+	meta := record.Metadata{}
+	// add original record's metadata to meta
+	for k, v := range rec.Metadata {
+		meta[k] = v
+	}
+
+	var structPayload record.StructuredData
+	if p, ok := rec.Payload.After.(record.StructuredData)["payload"].(map[string]interface{}); ok {
+		structPayload = p
+	} else {
+		return nil, cerrors.Errorf("payload doesn't contain a record")
+	}
```

Comment:
Nitpick: Line 302 is unreachable, if that would be the case the unwrapper would already throw an error in line 217. We could simplify.
```
structPayload := rec.Payload.After.(record.StructuredData)["payload"].(map[string]interface{})
```

Reply:
It just felt weird to ignore a casting error 😅 but yeah it shouldn't get here enyways

Reply:
You're right, maybe it would be nice to add a comment why we can safely do this.